runner.py

```diff
@@ -637,18 +638,51 @@ def setup_networks(self):
             self.__networks.append(network)
             self.__join_default_network = True
 
+    # Order service names based on 'depends_on'
+    def order_service_names(self, order_array, service_name, visited=None):
+        if visited is None:
+            visited = set()
+        if service_name in visited:
+            raise RuntimeError(f"Cycle found in depends_on definition with service '{service_name}'!")
+        visited.add(service_name)
+
+        service = self._usage_scenario['services'][service_name]
+
+        if 'depends_on' in service:
+            if isinstance(service['depends_on'], dict):
+                raise RuntimeError(f"Service definition of {service_name} uses the long form of 'depends_on', however, GMT only supports the short form!")
+            for dep in service['depends_on']:
+                if dep not in order_array:
+                    order_array = self.order_service_names(order_array, dep, visited)
+        
+        if service_name not in order_array:
+            order_array.append(service_name)
+
+        return order_array
+
     def setup_services(self):
+        config = GlobalConfig().config
+        print(TerminalColors.HEADER, '\nSetting up services', TerminalColors.ENDC)
         # technically the usage_scenario needs no services and can also operate on an empty list
         # This use case is when you have running containers on your host and want to benchmark some code running in them
-        for service_name in self._usage_scenario.get('services', []):
-            print(TerminalColors.HEADER, '\nSetting up containers', TerminalColors.ENDC)
+        services = self._usage_scenario.get('services', {})
+
+        order_of_service_names = []
+        # Check if there are dependencies defined with 'depends_on'.
+        # If so, change the order of the services accordingly.
+        for service_name in services.keys():
+            order_of_service_names = self.order_service_names(order_of_service_names, service_name)
```

Comment:
Hey, thank you for this great PR again. Arne and me just talked about the code and while this code is fully functional we see that it might be hard to maintain in the future as you carry the whole list `order_of_service_names` or `order_array` into every sub iteration. While this is fine for now we see that this might be the source of errors in a few years time as it is harder to comprehend. Also this pattern can be the source of errors when doing more complex things. When calling recursive functions it might be better to only pass in the context of the recursion and then handle the building of the list in the parent. Not something we need to fix right now but just as a remark.

Reply:
Based on this feedback I created an alternative PR with some refactorings:
https://github.com/green-coding-berlin/green-metrics-tool/pull/593